### ISSUE TITLE
Cache-control header

### ISSUE THREAD

**Issue by ariscn** [falconry/falcon#644]
`resp.cache_control = 'no-cache` splits the string into a list.

Falcon should support strings and lists for setting this header.

**Comment by philiptzou** [comment]
This is actually called "function overloading" which is commonly used in C and Java. But unlike C or Java, Python's approach of function overloading is a bit inefficient since you had to detect the value type in the function/method.

It's true function overloading is smart and friendly to programmers, but since falcon is dedicated to become a high-performance framework we can't bear with the inefficient comes with it. Also:

> In the face of ambiguity, refuse the temptation to guess.
> - The Zen of Python

**Comment by ariscn** [comment]
The framework can be performant and easy to use. Cecking `if isinstance(v, basestring): ....` is neither hard, bad, nor inefficient.

**Comment by MackYoel** [comment]
@philiptzou this reminds me of some methods of falcon. Personally I do not agree with that only lists accepted as iterables, tuples are faster to iterate that lists
<pre><code>
def prepare_middleware(middleware=None):
    """Check middleware interface and prepare it to iterate.

    Args:
        middleware:  list (or object) of input middleware

    Returns:
        A middleware list
    """

    # PERF(kgriffs): do getattr calls once, in advance, so we don't
    # have to do them every time in the request path.
    prepared_middleware = []

    if middleware is None:
        middleware = []
    else:
        if not isinstance(middleware, list):
            middleware = [middleware]

    for component in middleware:
        process_request = util.get_bound_method(component,
                                                'process_request')
        process_resource = util.get_bound_method(component,
                                                 'process_resource')
        process_response = util.get_bound_method(component,
                                                 'process_response')

        if not (process_request or process_resource or process_response):
            msg = '{0} does not implement the middleware interface'
            raise TypeError(msg.format(component))

        prepared_middleware.append((process_request, process_resource,
                                    process_response))

    return prepared_middleware
</code>
</pre>

**Comment by philiptzou** [comment]
@MackYoel True that tuples are faster than lists. but lists are mutable.

**Comment by MackYoel** [comment]
@philiptzou yes, but prepare_middleware is a pure function, it returns a new list, I think It should accept a iterable in this case. What do you think about it?

**Comment by philiptzou** [comment]
@MackYoel Yes I agree with you that the `prepare_middleware` is some kinda annoying... Personally I preferred to use `def prepare_middleware(*middlewares)` instead of detecting types in the function. Anyway, this discussion is not related to current subject of this page. So how about we start another new issue to discuss it further?

**Comment by philiptzou** [comment]
@ariscn But I think there're problems in the original design of `header_property` indeed. The main problem is, the `__get__()` should return the same type object which `__set__()` accepts. In the case of `cache_control`:

```python
resp.cache_control = ['no-cache']
assert resp.cache_control == 'no-cache'
```

I don't like it either. I believe a better API design should like this (What You Set is What You Get):

```python
resp.cache_control = ['no-cache']
assert resp.cache_control == ['no-cache']
```

Well this will be a breaking change.

**Comment by MackYoel** [comment]
@philiptzou I opened a very specific issue about it <a href="https://github.com/falconry/falcon/issues/630"></a> but nobody gave importance. Maybe you could open a more general issue

**Comment by MackYoel** [comment]
@philiptzou I opened a very specific issue about it [#630](https://github.com/falconry/falcon/issues/630) but nobody gave importance, maybe you could open a more general theme.

**Comment by kgriffs** [comment]
@philiptzou Would you mind breaking out your last suggestion re "What You Set is What You Get" to another issue so we can discuss it separately? 

As for whether these properties should support a single argument or a list of items, there are several interfaces in the framework that do this. So there would be some precedent for adding it in this case as well. However, I think that pattern may have been a mistake to follow, and overall it may be better to go modify all those places to just take an iterable, and only an iterable. Makes things more consistent, easier to reason about, etc. 

I could still be persuaded either way, but I think we need to decide which design direction to take as a community before working on this particular case involving `Response.cache_control`.
